types/aws-lambda/test/codebuild-tests.ts

```diff
@@ -0,0 +1,160 @@
+import { CodeBuildCloudWatchStateEvent } from 'aws-lambda';
+
+/* CodeBuild CloudWatch Events
+ * see https://docs.aws.amazon.com/codebuild/latest/userguide/sample-build-notifications.html
+ */
+const CodeBuildCloudWatchStateEvent: CodeBuildCloudWatchStateEvent = {
```

Comment:
Ideally tests should be using the handler types, like a user would, but we still have several in this style. It's much less important when there isn't a result type though.

Reply:
after I gave it a bit of thought it makes more sense to remove those Codebuild specific handlers. In my project I'm currently receiving these types as a message of `SNSEvent` triggered by CloudWatch event. That means that if I needed to create a handler, I would be just replicating `SNS` types
I already removed those handlers
Thanks

Reply:
Sorry, wrote the first review on a phone, so I was a bit terse. I'm referring to writing the test that uses the types as a user would use the types, e.g. writing a handler that reads the fields. This catches a lot of broken changes or unexpected warts in using types.

So for this package that would be writing a handler type, normally something like:

```ts
const someHandler: SomeHandler = async (event) => {
  strOrUndefined = event.someOptionalField;
  ...
};
```

though it's a bit more complicated with a result type.

Reply:
I understand what you mean about the test of handler. What I wanted to say is that handler does not directly receive `CodeBuildCloudWatchStateEvent`, but it receives `SNSEvent` with record that contains `CodeBuildCloudWatchStateEvent` as a JSON string in  `Message`

Something like
```
import { CodeBuildCloudWatchStateEvent, SNSEvent } from 'aws-lambda'

export const handleCodebuildEvent = async (event: SNSEvent) => {
  const record = event.Records[0]

  const codebuildEvent: CodeBuildCloudWatchStateEvent = JSON.parse(record.Sns.Message)
}
```

I haven't tried integration with EventBridge, only via CloudWatch Event and SNS.
Basically the use case is that you have CloudWatch event defined that checks CodeBuild state changes. Upon event it pushes a message to SNS topic and there is lambda subscribed to this SNS topic. This lambda consumes the messages  as in a code of block above.

I already have working monitoring of CodePipelines in the same way - CloudWatch event checks CodePipeline executions and upon event it pushes messages to SNS topic and there is lambda subscribed to that topic

```
import { CodePipelineCloudWatchActionEvent, SNSEvent } from 'aws-lambda'

export const handlePipelineEvent = async (event: SNSEvent) => {
  const record = event.Records[0]

  const pipelineEvent: CodePipelineCloudWatchActionEvent = JSON.parse(record.Sns.Message)
}
```

In this use case I'm not sure it makes sense to create a test for handler, cause I would be testing the `SNSEvent` in the handler, not the `CodeBuildCloudWatchStateEvent` - since handler receives type of `SNSEvent` and the `CodeBuildCloudWatchStateEvent` is just a stringified JSON in the `SNSEvent`
However if you feel strong about it I can add the test that will contain `SNSEvent` as an event type of handler and it will contain JSON string of `CodeBuildCloudWatchStateEvent` and in the end I will just check whether the `Message` of `SNSEvent` consists the same stringified JSON

Reply:
EventBridge is just CloudWatch Events renamed. Presumably you can just send these events straight to lambda: (I haven't messed with anything other than scheduled events), this would be the handler type.

I would *somewhat* strongly prefer that the test was reading from the event, not creating one, as that catches problems with changes to the structure.

With SNS usage as the test, this might look something like:

```ts
export const handlePipelineEvent = async (snsEvent: SNSEvent) => {
  const event: CodePipelineCloudWatchActionEvent = JSON.parse(snsEvent.Records[0].Sns.Message)

  str = event.version;
  str = event.id;
  // and so on for the rest of the fields, using types from `aws-lambda-tests.ts` where appropriate.
}
```

but obviously, that's quite a decent bit of work, so I'm not going to block on that, much better to have this merged.

Reply:
For comparison, here's an example of a test parsing data from the event to get further data:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/aws-lambda/test/cloudwatch-tests.ts#L13-L15

Note that unless you're getting this as an event directly in a handler, this type shouldn't have the `...Event` suffix, maybe something like `...MessageData`? But I think you can and thus should use the EventBridge handler structure.

Reply:
ok, no problem
I'll modify it and add the tests 👍